### PR TITLE
[WIP] Enable frontend preview for paid blocks requiring upgrade

### PR DESCRIPTION
**Pull request by stacimc**
Fixes 158-gh-Automattic/view-design and 157-gh-Automattic/view-design and 175-gh-Automattic/view-design.

#### Changes proposed in this Pull Request:

This PR re-enables frontend previews for paid blocks requiring an upgrade. If a block is unavailable due to not meeting plan requirements _and the current user is an admin_, the block will still be rendered on the frontend with an upgrade banner identical to the one that appears in the editor.
 
Screen Shot 2020-12-18 at 4 11 30 PM

It also refactors the existing UpgradePlanBanner used in the editor to pull out as much common code as possible into a generic Nudge component that can be server-side-rendered.

#### Jetpack product discussion

Internal reference: pcjTuq-8U-p2

#### Does this pull request change what data or activity we track or use?

* TBD

#### Testing instructions:

This requires re-testing across all the scenarios: wpcom and Jetpack, with and without the required plan, and with and without Stripe connected.

Follow testing instructions on #17907.

Note: the wp-env scripts will not sync changes to `class.jetpack-gutenberg.php` or `class-blocks.php`. These may need to be manually synced to their respective files in the sandbox.

In order to test on free Jetpack sites, you will need to turn on the `jetpack_block_editor_enable_upgrade_nudge` filter locally.

#### Proposed changelog entry for your changes:

* TBD

### CODE REVIEW

File: extensions/blocks/premium-content/premium-content.php [hunk 1]
```diff
@@ -32,6 +31,7 @@ function register_block() {
 		FEATURE_NAME,
 		array(
 			'render_callback' => __NAMESPACE__ . '\render_block',
+			'enable_preview'  => true,
```
Comment (stacimc) [inline extensions/blocks/premium-content/premium-content.php:34]: Taking suggestions for a better name for this option.

File: class.jetpack-gutenberg.php [hunk 2]
```diff
@@ -922,6 +922,30 @@ public static function validate_block_embed_url( $url, $allowed = array(), $is_r
...
+		$availability = self::get_availability();
+
+		return (
+			isset( $availability[ $slug ]['details']['required_plan'] ) &&
+			current_user_can( 'manage_options' )
+			/** This filter is documented in class.jetpack-gutenberg.php */
+			&& apply_filters( 'jetpack_block_editor_enable_upgrade_nudge', false )
+			/** This filter is documented in _inc/lib/admin-pages/class.jetpack-react-page.php */
+			&& apply_filters( 'jetpack_show_promotions', true )
+			&& ! is_feed()
```
Comment (stacimc) [inline class.jetpack-gutenberg.php:945]: The last two conditions here were part of the original check for rendering a frontend upgrade nudge. I am looking for more context to see if they are still needed, particularly the `jetpack_show_promotions` filter.

Comment (apeatling) [reply class.jetpack-gutenberg.php:945]: I'm not seeing any usage of the `jetpack_show_promotions` filter across the Jetpack or WordPress.com codebase. I suspect this is a global way to turn off all promotional notifications across Jetpack. It's possible that this is being used by hosts somewhere though, so best to leave it in.

Comment (apeatling) [reply class.jetpack-gutenberg.php:945]: `jetpack_block_editor_enable_upgrade_nudge` is the current way we enable the display of upgrade nudges on WordPress.com. Do we still need this with this work?

File: extensions/blocks/premium-content/premium-content.php [hunk 3]
```diff
@@ -61,40 +61,21 @@ function render_block( $attributes, $content ) {
...
+		&& ! membership_checks()
 	) {
-		$upgrade_nudge = render_upgrade_nudge();
-		return $upgrade_nudge . $content;
-	}
-
-	// Stripe connection nudge.
-	if ( ! membership_checks() && current_user_can_edit() ) {
-		$stripe_nudge = render_stripe_nudge();
+		$stripe_nudge = render_stripe_nudge;
```
Comment (apeatling) [inline extensions/blocks/premium-content/premium-content.php:71]: Missing `()`.

File: extensions/blocks/premium-content/premium-content.php (hunk 1 above)
Comment (apeatling) [reply extensions/blocks/premium-content/premium-content.php:34]: `enable_frontend_preview` might be slightly more explanatory?

**Comment by stacimc** [comment]
> jetpack_block_editor_enable_upgrade_nudge is the current way we enable the display of upgrade nudges on WordPress.com. Do we still need this with this work?

This enables upgrade nudges in the editor, yes. It was part of the original conditional for displaying frontend upgrade nudges as well, so I included it in the first pass. I think you're right and it can be removed -- my thinking is:

If upgrade nudges in the editor weren't enabled, the user wouldn't have been able to insert the block at all, so we should never end up in a scenario where we're rendering a paid block on the frontend of a free site that does not support nudges.

**Comment by jeherve** [comment]
Somewhat related: #14893, #18054. 

Internal reference: pcjTuq-8U-p2

**Comment by stacimc** [comment]
Closing in favor of #18239

**Pull request closed without merging**